lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,817 @@
+	pub fn paths(&self) -> &[BlindedPath] {
+		self.contents.paths.as_ref().map(|paths| paths.as_slice()).unwrap_or(&[])
+	}
+
+	/// The minimum quantity of items supported.
+	pub fn quantity_min(&self) -> u64 {
+		self.contents.quantity_min()
+	}
+
+	/// The maximum quantity of items supported.
+	pub fn quantity_max(&self) -> u64 {
+		self.contents.quantity_max()
+	}
+
+	/// The public key used by the recipient to sign invoices.
+	pub fn signing_pubkey(&self) -> PublicKey {
+		self.contents.signing_pubkey.unwrap()
+	}
+
+	#[cfg(test)]
+	fn as_tlv_stream(&self) -> OfferTlvStreamRef {
+		self.contents.as_tlv_stream()
+	}
+}
+
+impl OfferContents {
+	pub fn amount_msats(&self) -> u64 {
+		self.amount.as_ref().map(Amount::as_msats).unwrap_or(0)
+	}
+
+	pub fn quantity_min(&self) -> u64 {
+		self.quantity_min.unwrap_or(1)
+	}
+
+	pub fn quantity_max(&self) -> u64 {
+		self.quantity_max.unwrap_or_else(||
+			self.quantity_min.map_or(1, |_| u64::max_value()))
+	}
+
+	fn as_tlv_stream(&self) -> OfferTlvStreamRef {
+		let (currency, amount) = match &self.amount {
+			None => (None, None),
+			Some(Amount::Bitcoin { amount_msats }) => (None, Some(*amount_msats)),
+			Some(Amount::Currency { iso4217_code, amount }) => (
+				Some(iso4217_code), Some(*amount)
+			),
+		};
+
+		let features = {
+			if self.features == OfferFeatures::empty() { None } else { Some(&self.features) }
+		};
+
+		OfferTlvStreamRef {
+			chains: self.chains.as_ref(),
+			metadata: self.metadata.as_ref(),
+			currency,
+			amount,
+			description: Some(&self.description),
+			features,
+			absolute_expiry: self.absolute_expiry.map(|duration| duration.as_secs()),
+			paths: self.paths.as_ref(),
+			issuer: self.issuer.as_ref(),
+			quantity_min: self.quantity_min,
+			quantity_max: self.quantity_max,
+			node_id: self.signing_pubkey.as_ref(),
+		}
+	}
+}
+
+impl Writeable for OfferContents {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		self.as_tlv_stream().write(writer)
+	}
+}
+
+/// The minimum amount required for an item in an [`Offer`], denominated in either bitcoin or
+/// another currency.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Amount {
+	/// An amount of bitcoin.
+	Bitcoin {
+		/// The amount in millisatoshi.
+		amount_msats: u64,
+	},
+	/// An amount of currency specified using ISO 4712.
+	Currency {
+		/// The currency that the amount is denominated in.
+		iso4217_code: CurrencyCode,
+		/// The amount in the currency unit adjusted by the ISO 4712 exponent (e.g., USD cents).
+		amount: u64,
+	},
+}
+
+impl Amount {
+	/// Returns the amount in millisatoshi.
+	pub fn as_msats(&self) -> u64 {
+		match self {
+			Amount::Currency { .. } => unimplemented!(),
+			Amount::Bitcoin { amount_msats } => *amount_msats,
+		}
+	}
+}
+
+/// An ISO 4712 three-letter currency code (e.g., USD).
+pub type CurrencyCode = [u8; 3];
+
+tlv_stream!(OfferTlvStream, OfferTlvStreamRef, {
+	(2, chains: Vec<ChainHash>),
```

Comment:
It depends on the context, though note that all the records use `Option` in  the TLV structs.

For `Offer`, only `description` and `node_id` are required. This is currently enforced at compile-time by `OfferBuilder::new` requiring both as parameters.

However, for `InvoiceRequest` used as a "send invoice" (e.g., refund), `node_id` MUST NOT be set.

I also avoided the `lightning-invoice`-style compile time checks because IIUC it can't be easily done in bindings since each builder function returns a different type, essentially.